### ISSUE TITLE
Error: Missing signing file for certificate

### ISSUE THREAD

**Issue by BrockCruess** [vacuumlabs/cardano-hw-cli#183]
```
cardano-hw-cli transaction witness \
  --tx-file tx.raw \
  --hw-signing-file hw-stake.hwsfile \
  --mainnet \
  --out-file hw-stake.witness

Error: The requested operation failed. Check that your Ledger device is connected, unlocked and with Cardano app running.
Details: Error: Missing signing file for certificate
```

Using 1.16 as the latest main release 1.14 doesn't work with latest Ledger Nano X Cardano app.

Ledger is definitely on with Cardano app ready and waiting.
hw-stake.hwsfile is definitely present where the command is being run from, so not sure what signing file could be missing.

**Comment by davidmisiak** [comment]
Could you please provide example contents of `tx.raw` and `hw-stake.hwsfile` that trigger the issue?

**Comment by BrockCruess** [comment]
> Could you please provide example contents of `tx.raw` and `hw-stake.hwsfile` that trigger the issue?

tx.raw:
```
{
    "type": "Unwitnessed Tx BabbageEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a500818258208c25d6f9064747b62a14fc271d7e255c7f95ac54eff13ec268056401559df8c20001818258390108979b5e933b166645656501a17720b4e4d2d1c53ea9be446dd917fb96af8d3fba7c34af38f1caf9f576ee798abfb81bfe944ca4f42348681a0034f005021a0002f3f9031a0898f66c04818a03581c0df0f6e5a3191520aa0a58268c38fe608d1d931766fc006635f3f2b15820a2d3c275f518a126e9d5d471ee4958b3fbe35bc92ff8a27d8ff4a8603ddd32ef1b00000004a817c8001a0a21fe80d81e820001581de1f2eece9ca98e3bd1407a0181fbc6a26dfbe6b7f7ad759579389da3b581581cf2eece9ca98e3bd1407a0181fbc6a26dfbe6b7f7ad759579389da3b58183011917707262726f636b706f6f6c2e64646e732e6e657482781c68747470733a2f2f74696e7975726c2e636f6d2f3239657a367966365820a36a183e3b8ca6cea4e894c5e1c9c97bbc61c2c11bdb180c1d040657910908d8a0f5f6"
}
```

tx.raw was created by this command, after tx_in, txOut, currentSlot and fee were defined:
```
cardano-cli transaction build-raw \
    ${tx_in} \
    --tx-out $(cat payment.addr)+${txOut} \
    --invalid-hereafter $(( ${currentSlot} + 9999999)) \
    --fee ${fee} \
    --certificate-file pool-registration.cert \
    --out-file tx.raw
```

I'm not too keen on sharing hw-stake.hwsfile as I don't personally know the extents of its contents but it was generated successfully with this command:
```
cardano-hw-cli address key-gen \
  --path 1852H/1815H/0H/2/0 \
  --verification-key-file hw-stake.vkey \
  --hw-signing-file hw-stake.hwsfile
```

I've been using this method with a Nano X and cardano-hw-cli for a long time and never had issues until recently so I'm confident I'm doing everything right. I'm trying to rotate ownership of my stake pool to a new Ledger Nano X.

**Comment by davidmisiak** [comment]
That is quite weird, we haven't modified the pool registration code in quite a long time, as far as I know.

When you run `cardano-cli stake-address key-hash --stake-verification-key-file hw-stake.vkey`, do you get `f2eece9ca98e3bd1407a0181fbc6a26dfbe6b7f7ad759579389da3b5` (the pool owner stake key present in the tx CBOR)?

**Comment by BrockCruess** [comment]
No, I get the key-hash of the hardware wallet, which isn't the pool owner yet. It is the correct key-hash of the hardware wallet though, I can confirm.

**Comment by davidmisiak** [comment]
When you are signing a pool registration tx, cardano-hw-cli determines the signing mode (whether you're signing as the pool operator or as a pool owner) based on the provided hwsfiles. You haven't provided the pool key hwsfile nor a payment hwsfile, so cardano-hw-cli sets the POOL_REGISTRATION_AS_OWNER signing mode. And for that, you need to provide the pool owner staking key hwsfile.

[This document](https://github.com/vacuumlabs/cardano-hw-cli/blob/develop/docs/pool-registration.md) explains some more details.

**Comment by gitmachtl** [comment]
I did a quick test here, added a new ledger pool owner to my existing sanchonet pool. I had no issues at all.

Cardano-HW-CLI: v1.16.0
Ledger-Cardano-App: v7.1.3

Here is the pool registration/update certificate:
```json
{
    "type": "CertificateConway",
    "description": "Stake Pool Registration Certificate",
    "cborHex": "8a03581c1b7a61ba3e5292f34666af3bb2f0b8d36e5001aa41e235709ffa69105820fdb9ec5d3e9774458dbdb7b337190fc9a2fbc585f28060fd0158ee8f9a96b8a2001b00000002540be400d81e82010a581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fd9010282581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf581cc13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73f818301190bb971746573742e7374616b65706f6f6c2e617482782e68747470733a2f2f6d792d69702e61742f746573742f73616e63686f706f6f6c2e6d657461646174612e6a736f6e582069059846a028ab1161218d3b683502abcc0513a88567de78cc18b43d506e3f3c"
}
```

Here is the unwitnessed tx body:
```json
{
    "type": "Unwitnessed Tx ConwayEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a500d9010281825820d750f9cb4d3a8eba7a57604cc97a7453f340c92a3bcaa3db9b89fc2dcbbb80fb00018182581d6052e63f22c5107ed776b70f7b92248b02552fd08f3e747bc7450994411b00000022cf34aaa0021a0002e7f1031a02551ea304d90102818a03581c1b7a61ba3e5292f34666af3bb2f0b8d36e5001aa41e235709ffa69105820fdb9ec5d3e9774458dbdb7b337190fc9a2fbc585f28060fd0158ee8f9a96b8a2001b00000002540be400d81e82010a581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fd9010282581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf581cc13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73f818301190bb971746573742e7374616b65706f6f6c2e617482782e68747470733a2f2f6d792d69702e61742f746573742f73616e63686f706f6f6c2e6d657461646174612e6a736f6e582069059846a028ab1161218d3b683502abcc0513a88567de78cc18b43d506e3f3ca0f5f6"
}
```

And here is the witnessed one:
```json
{
    "type": "Witnessed Tx ConwayEra",
    "description": "Ledger Cddl Format",
    "cborHex": "84a500d9010281825820d750f9cb4d3a8eba7a57604cc97a7453f340c92a3bcaa3db9b89fc2dcbbb80fb00018182581d6052e63f22c5107ed776b70f7b92248b02552fd08f3e747bc7450994411b00000022cf34aaa0021a0002e7f1031a02551ea304d90102818a03581c1b7a61ba3e5292f34666af3bb2f0b8d36e5001aa41e235709ffa69105820fdb9ec5d3e9774458dbdb7b337190fc9a2fbc585f28060fd0158ee8f9a96b8a2001b00000002540be400d81e82010a581de0c13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73fd9010282581c03d205532089ad2f7816892e2ef42849b7b52788e41b3fd43a6e01cf581cc13582aec9a44fcc6d984be003c5058c660e1d2ff1370fd8b49ba73f818301190bb971746573742e7374616b65706f6f6c2e617482782e68747470733a2f2f6d792d69702e61742f746573742f73616e63686f706f6f6c2e6d657461646174612e6a736f6e582069059846a028ab1161218d3b683502abcc0513a88567de78cc18b43d506e3f3ca100d9010284825820c13745be35c2dfc3fa9523140030dda5b5346634e405662b1aae5c61389c55b35840a691c4dcd9065240e0da51cb3b70c171a87726ccb9c2e9d65e892b206b2f7dbbdadbaff10715ab6121626e3ec9d08cdc8507fa2238ac379c316f361d61473f0a8258200ce27374a90797cefa936725913ba4a1b58bb9a78db38d896c4c1c4663ec589c5840b34bf5a6f87b9c05f778039764f5b49ba0fa36eeb9149d2b92f3a195a5b367d773fcf6d55f2d73a68ae390acbfbd0e68cc3102c1b234fe73adbe04ab574d9b09825820742d8af3543349b5b18f3cba28f23b2d6e465b9c136c42e1fae6b2390f565427584001799b655794c537f8b88c1cfee482e140ba5f43b219e3f2cd876c95c511e1d80f2db4fb3c48ce77996272a0405208f5f32a8670fb656e7d733164ab4829c0038258209be513df12b3fabe7c1b8c3f9fab0968eb2168d5689bf981c2f7c35b11718b2758401ca969bbaca04e84ea17f737a7238f2cf1821e3a2e8aa23268a9868fea9d701b955b560fac3e117efb6a4d66e24ad6c5767caebad580811f381c7cf9b480030df5f6"
}
```

The hwsfile is:
```json
{
    "type": "StakeHWSigningFileShelley_ed25519",
    "description": "Stake Hardware Signing File",
    "path": "1852H/1815H/0H/2/0",
    "cborXPubKeyHex": "5840c13745be35c2dfc3fa9523140030dda5b5346634e405662b1aae5c61389c55b30d113368e4f448258a295e3ede8d28953bf406ef2073a09e7343bdc0bb7999e6"
}
```

Transaction was paid via a normal CLI-Wallet.

**Comment by BrockCruess** [comment]
The pool key and payment key can't be hwsfile because they're not hardware keys. The pool .skey and payment .skey are used to multisign the transaction at the end, along with all other required signatures.

I'm trying to _add_ a hardware wallet as a pool owner, and make it the rewards account. hw-stake.vkey is what's new here, owner-stake.vkey is what existed already.

Here's an example of the registration cert command, paths/filenames modified for privacy:
```
cardano-cli stake-pool registration-certificate \
    --pool-pledge [PLEDGE] \
    --pool-cost 170000000 \
    --pool-margin 0 \
    --single-host-pool-relay [RELAY] \
    --pool-relay-port [PORT] \
    --metadata-url [URL] \
    --metadata-hash $(cat poolMetadataHash.txt) \
    --cold-verification-key-file pool.vkey \
    --vrf-verification-key-file vrf.vkey \
    --pool-reward-account-verification-key-file hw-stake.vkey \
    --pool-owner-stake-verification-key-file hw-stake.vkey \
    --pool-owner-stake-verification-key-file owner-stake.vkey \
    --mainnet \
    --out-file pool-registration.cert
```

This works fine, the registration cert isn't the issue I'm facing. I need a hardware witness file to sign the transaction with, that's where the issue comes in:
```
cardano-hw-cli transaction witness \
  --tx-file tx.raw \
  --hw-signing-file hw-stake.hwsfile \
  --mainnet \
  --out-file hw-stake.witness
```

I would then go on to multisign the tx to submit the cert, using:
```
cardano-cli transaction assemble \
  --tx-body-file $HOME/cardano/tx.raw \
  --witness-file $HOME/cardano/hw-stake.witness \
  --witness-file $HOME/cardano/pool.witness \
  --witness-file $HOME/cardano/payment.witness \
  --witness-file $HOME/cardano/stake.witness \
  --witness-file $HOME/cardano/owner.witness \
  --out-file $HOME/cardano/tx.multisign
```
But I can't get to this point without hw-stake.witness

**Comment by gitmachtl** [comment]
Not sure what you're doing wrong, because i generate the witness from a hw-wallet pretty similar in my scripts with those lines:
https://github.com/gitmachtl/scripts/blob/3c47474b89cbd454844804c08ce14171e425f0cb/cardano/mainnet/05c_regStakepoolCert.sh#L811-L821

Also, make sure to always do a autocorrect of the tx before you further process like:
https://github.com/gitmachtl/scripts/blob/3c47474b89cbd454844804c08ce14171e425f0cb/cardano/mainnet/00_common.sh#L2606

**Comment by davidmisiak** [comment]
> Here's an example of the registration cert command

I don't think that a certificate generated by this command corresponds to the one in the tx CBOR that you posted earlier: The tx CBOR contains only a single pool owner stake key - the same one like the pool rewards stake key (the one with hash `f2eece9ca98e3bd1407a0181fbc6a26dfbe6b7f7ad759579389da3b5`).

**Comment by gitmachtl** [comment]
Just to make it identical, i tried it also now again like in your setup @BrockCruess with a local cli owner, a hw owner and the same hw as the rewards account. Same result, no issues. Also just retried it again with all babbage-era settings and not conway, same outcome.

**Comment by BrockCruess** [comment]
I went through the whole process start to finish again and it worked this time. Very odd, as I don't think I changed anything. Maybe an old registration certificate was stuck and couldn't be overwritten when I re-ran my commands, hence why the registration cert didn't look like what I was describing? Not sure.
Thanks for sitting through this with me, looking forward to Ledger Stax support being added by the way!
